File: 06-DevOps/01-容器化/OCI分发协议与镜像拉取/go/dist_registry.go
// OCI Distribution Spec 的 registry 侧模型（Python 模型的 Go 复刻）。
//
// 语义对照 opencontainers/distribution-spec spec.md：
//
//	§Endpoints（end-1 .. end-14）、§Error Codes（code-1 .. code-14）
//	§Pushing blobs monolithically / in chunks、§Mounting、§Listing Tags
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"regexp"
	"sort"
	"strings"
)

var contentRangeRe = regexp.MustCompile(`^[0-9]+-[0-9]+$`)

// RegistryError 对应规范 §Error Codes 的 JSON 错误体。
type RegistryError struct {
	Status  int
	Code    string
	Message string
	Detail  map[string]string
}

func (e *RegistryError) Error() string { return e.Code }

func (e *RegistryError) Body() map[string]interface{} {
	err := map[string]interface{}{"code": e.Code, "message": e.Message}
	if e.Detail != nil {
		err["detail"] = e.Detail
	}
	return map[string]interface{}{"errors": []interface{}{err}}
}

func regErr(status int, code, msg string) *RegistryError {
	return &RegistryError{Status: status, Code: code, Message: msg}
}

// Response 是最小 HTTP 响应模型。
type Response struct {
	Status  int
	Headers map[string]string
	Body    interface{}
}

func digestOf(data []byte) string {
	sum := sha256.Sum256(data)
	return "sha256:" + hex.EncodeToString(sum[:])
}

type UploadSession struct {
	UUID string
	Name string
	Data []byte
}

func (s *UploadSession) Offset() int { return len(s.Data) }

type Repo struct {
	Manifests map[string]map[string]interface{}
	Tags      map[string]string
	Blobs     map[string][]byte
}

type Registry struct {
	Repos      map[string]*Repo
	Uploads    map[string]*UploadSession
	Delete     bool
	nextUUID   int
}

func NewRegistry(deleteEnabled bool) *Registry {
	return &Registry{Repos: map[string]*Repo{}, Uploads: map[string]*UploadSession{},
		Delete: deleteEnabled}
}

func (rg *Registry) EnsureRepo(name string) *Repo {
	if _, ok := rg.Repos[name]; !ok {
		rg.Repos[name] = &Repo{Manifests: map[string]map[string]interface{}{},
			Tags: map[string]string{}, Blobs: map[string][]byte{}}
	}
	return rg.Repos[name]
}

// Repo 对应 end-1..end-12 共用的仓库定位：不存在的仓库一律 404。
func (rg *Registry) Repo(name string) (*Repo, *RegistryError) {
	if r, ok := rg.Repos[name]; ok {
		return r, nil
	}
	return nil, regErr(404, "NAME_UNKNOWN", "repository name not known to registry")
}

func (rg *Registry) GetV2() *Response { return &Response{Status: 200, Headers: map[string]string{}} }

// end-2：GET /v2/<name>/blobs/<digest>
func (rg *Registry) GetBlob(name, digest string) (*Response, *RegistryError) {
	r, err := rg.Repo(name)
	if err != nil {
		return nil, err
	}
	data, ok := r.Blobs[digest]
	if !ok {
		return nil, regErr(404, "BLOB_UNKNOWN", "blob unknown to registry")
	}
	return &Response{Status: 200, Headers: map[string]string{
		"Docker-Content-Digest": digest, "Content-Length": fmt.Sprint(len(data))},
		Body: data}, nil
}

// end-4a/4b：POST /v2/<name>/blobs/uploads/
func (rg *Registry) PostUpload(name, mount, from string) (*Response, *RegistryError) {
	if mount != "" && from != "" {
		if src, ok := rg.Repos[from]; ok {
			if blob, ok := src.Blobs[mount]; ok {
				rg.EnsureRepo(name).Blobs[mount] = blob
				return &Response{Status: 201, Headers: map[string]string{
					"Location": "/v2/" + name + "/blobs/" + mount}}, nil
			}
		}
	}
	rg.nextUUID++
	s := &UploadSession{UUID: fmt.Sprintf("u%d", rg.nextUUID), Name: name}
	rg.Uploads[s.UUID] = s
	return &Response{Status: 202, Headers: map[string]string{
		"Location": "/v2/" + name + "/blobs/uploads/" + s.UUID,
		"Docker-Upload-UUID": s.UUID, "Range": "0-0"}}, nil
}

func (rg *Registry) session(uuid string) (*UploadSession, *RegistryError) {
	if s, ok := rg.Uploads[uuid]; ok {
		return s, nil
	}
	return nil, regErr(404, "BLOB_UPLOAD_UNKNOWN", "blob upload unknown to registry")
}

// end-5：PATCH，range 必须连续且首个必须从 0 起。
func (rg *Registry) PatchUpload(uuid string, data []byte, contentRange string) (*Response, *RegistryError) {
	s, err := rg.session(uuid)
	if err != nil {
		return nil, err
	}
	if contentRange != "" {
		if !contentRangeRe.MatchString(contentRange) {
			return nil, regErr(416, "BLOB_UPLOAD_INVALID", "bad content range")
		}
		parts := strings.SplitN(contentRange, "-", 2)
		start := atoi(parts[0])
		if s.Offset() == 0 && start != 0 {
			return nil, regErr(416, "BLOB_UPLOAD_INVALID",
				"the first chunk's range must begin with 0")
		}
		if start != s.Offset() {
			return nil, regErr(416, "BLOB_UPLOAD_INVALID", "chunk out of order")
		}
		if len(data) != atoi(parts[1])-start+1 {
			return nil, regErr(416, "BLOB_UPLOAD_INVALID",
				"content-length does not match range")
		}
	}
	s.Data = append(s.Data, data...)
	return &Response{Status: 202, Headers: map[string]string{
		"Range": fmt.Sprintf("0-%d", s.Offset()-1),
		"Docker-Upload-UUID": s.UUID}}, nil
}

// end-6：PUT ?digest=<digest> 关闭会话，digest 不匹配即 DIGEST_INVALID。
func (rg *Registry) PutUpload(uuid, digest string, data []byte, contentRange string) (*Response, *RegistryError) {
	s, err := rg.session(uuid)
	if err != nil {
		return nil, err
	}
	if len(data) > 0 {
		if _, err := rg.PatchUpload(uuid, data, contentRange); err != nil {
			return nil, err
		}
	}
	got := digestOf(s.Data)
	if got != digest {
		return nil, &RegistryError{Status: 400, Code: "DIGEST_INVALID",
			Message: "provided digest did not match uploaded content",
			Detail:  map[string]string{"provided": digest, "computed": got}}
	}
	rg.EnsureRepo(s.Name).Blobs[got] = s.Data
	delete(rg.Uploads, uuid)
	return &Response{Status: 201, Headers: map[string]string{
		"Location": "/v2/" + s.Name + "/blobs/" + got,
		"Docker-Content-Digest": got}}, nil
}

func (rg *Registry) DeleteUpload(uuid string) (*Response, *RegistryError) {
	if _, err := rg.session(uuid); err != nil {
		return nil, err
	}
	delete(rg.Uploads, uuid)
	return &Response{Status: 204}, nil
}

// end-8：ASCIIbetical 排序 + n/last 分页，last 非包含。
func (rg *Registry) ListTags(name string, n int, last string) (*Response, *RegistryError) {
	r, err := rg.Repo(name)
	if err != nil {
		return nil, err
	}
	tags := make([]string, 0, len(r.Tags))
	for t := range r.Tags {
		tags = append(tags, t)
	}
	sort.Strings(tags)
	if last != "" {
		kept := tags[:0]
		for _, t := range tags {
			if t > last {
				kept = append(kept, t)
			}
		}
		tags = kept
	}
	headers := map[string]string{}
	page := tags
	if n > 0 {
		if n < len(tags) {
			page = tags[:n]
			headers["Link"] = fmt.Sprintf("</v2/%s/tags/list?n=%d&last=%s>; rel=\"next\"",
				name, len(tags)-n, page[len(page)-1])
		}
	} else if n == 0 {
		page = nil
	}
	return &Response{Status: 200, Headers: headers,
		Body: map[string]interface{}{"name": name, "tags": page}}, nil
}

// end-10：DELETE /v2/<name>/blobs/<digest>
func (rg *Registry) DeleteBlob(name, digest string) (*Response, *RegistryError) {
	r, err := rg.Repo(name)
	if err != nil {
		return nil, err
	}
	if !rg.Delete {
		return nil, regErr(405, "UNSUPPORTED", "the operation is unsupported")
	}
	if _, ok := r.Blobs[digest]; !ok {
		return nil, regErr(404, "BLOB_UNKNOWN", "blob unknown to registry")
	}
	delete(r.Blobs, digest)
	return &Response{Status: 202}, nil
}

func atoi(s string) int {
	n := 0
	fmt.Sscanf(s, "%d", &n)
	return n
}

File: 06-DevOps/01-容器化/OCI分发协议与镜像拉取/python/dist_registry.py
"""OCI Distribution Spec 的 registry 侧模型。

逐节转写自 opencontainers/distribution-spec `spec.md`（54228 B，main 分支）：
  §Endpoints（end-1 .. end-14）、§Error Codes（code-1 .. code-14）、
  §Pushing blobs monolithically / in chunks、§Mounting、§Listing Tags、§Listing Referrers。
所有状态码、头部名与错误码都照抄规范，未做"我觉得更合理"的改动。
"""

import hashlib
import json
import re
import uuid

CONTENT_RANGE_RE = re.compile(r"^[0-9]+-[0-9]+$")

MANIFEST_MT = "application/vnd.oci.image.manifest.v1+json"
INDEX_MT = "application/vnd.oci.image.index.v1+json"


class RegistryError(Exception):
    """规范 §Error Codes 的 JSON 错误体。"""

    def __init__(self, status, code, message="", detail=None):
        super().__init__(code)
        self.status = status
        self.code = code
        self.message = message
        self.detail = detail

    def body(self):
        err = {"code": self.code, "message": self.message}
        if self.detail is not None:
            err["detail"] = self.detail
        return {"errors": [err]}


class Response:
    def __init__(self, status, headers=None, body=None):
        self.status = status
        self.headers = headers or {}
        self.body = body


def digest_of(data):
    return "sha256:" + hashlib.sha256(data).hexdigest()


class UploadSession:
    def __init__(self, name):
        self.uuid = str(uuid.uuid4())
        self.name = name
        self.data = b""

    @property
    def offset(self):
        return len(self.data)


class Registry:
    def __init__(self, delete_enabled=True, referrers_enabled=False):
        self.repos = {}                 # name -> repo dict
        self.uploads = {}               # uuid -> UploadSession
        self.delete_enabled = delete_enabled
        self.referrers_enabled = referrers_enabled

    # ---------------------------------------------------------------- 基础设施
    def repo(self, name):
        if not name or not re.match(r"^[a-z0-9]+(?:(?:[._]|__|[-]*)[a-z0-9]+)*$", name):
            raise RegistryError(404, "NAME_INVALID", "invalid repository name")
        if name not in self.repos:
            raise RegistryError(404, "NAME_UNKNOWN", "repository name not known to registry")
        return self.repos[name]

    def ensure_repo(self, name):
        if name not in self.repos:
            self.repos[name] = {"manifests": {}, "tags": {}, "blobs": {},
                                "referrers": {}}
        return self.repos[name]

    def _check_refs(self, repo, manifest):
        """MANIFEST_BLOB_UNKNOWN：manifest 引用的层/配置必须在仓库里。"""
        missing = []
        for ref in [manifest.get("config", {}).get("digest")] + \
                [l.get("digest") for l in manifest.get("layers", [])]:
            if ref and ref not in repo["blobs"]:
                missing.append(ref)
        if missing:
            raise RegistryError(404, "MANIFEST_BLOB_UNKNOWN",
                                "manifest references unknown blob",
                                {"digests": missing})

    # ---------------------------------------------------------------- end-1
    def get_v2(self):
        return Response(200, {}, {})

    # ---------------------------------------------------------------- end-2/3
    def get_blob(self, name, digest):
        repo = self.repo(name)
        if digest not in repo["blobs"]:
            raise RegistryError(404, "BLOB_UNKNOWN", "blob unknown to registry")
        data = repo["blobs"][digest]
        return Response(200, {"Docker-Content-Digest": digest,
                              "Content-Length": str(len(data))}, data)

    def head_blob(self, name, digest):
        r = self.get_blob(name, digest)
        return Response(200, r.headers, b"")

    def get_manifest(self, name, reference):
        repo = self.repo(name)
        digest = repo["tags"].get(reference, reference)
        if digest not in repo["manifests"]:
            raise RegistryError(404, "MANIFEST_UNKNOWN",
                                "manifest unknown to registry")
        body = repo["manifests"][digest]
        return Response(200, {"Docker-Content-Digest": digest,
                              "Content-Type": body["mediaType"]}, body)

    def head_manifest(self, name, reference):
        r = self.get_manifest(name, reference)
        return Response(200, r.headers, None)

    # ---------------------------------------------------------------- end-7
    def put_manifest(self, name, reference, manifest, media_type=MANIFEST_MT,
                     extra_tags=None):
        repo = self.ensure_repo(name)
        self._check_refs(repo, manifest)
        raw = json.dumps(manifest, sort_keys=True).encode()
        digest = digest_of(raw)
        repo["manifests"][digest] = dict(manifest, mediaType=media_type)
        refs = [reference] + list(extra_tags or [])
        for ref in refs:
            repo["tags"][ref] = digest
        if manifest.get("subject"):
            repo["referrers"].setdefault(manifest["subject"], []).append(digest)
        return Response(201, {"Location": "/v2/%s/manifests/%s" % (name, digest),
                              "Docker-Content-Digest": digest}, None)

    # ---------------------------------------------------------------- end-4/5/6
    def post_upload(self, name, digest=None, mount=None, frm=None,
                    content_length=0):
        if mount and frm:
            src = self.repo(frm) if frm in self.repos else None
            if src and mount in src["blobs"]:
                dst = self.ensure_repo(name)
                dst["blobs"][mount] = src["blobs"][mount]
                return Response(201, {"Location": "/v2/%s/blobs/%s" % (name, mount)})
            # 挂载失败：回落到普通上传流程（仍是 202）
        if digest is not None:
            # end-4b：单次 POST 整体上传，digest 提前给出
            s = UploadSession(name)
            self.uploads[s.uuid] = s
            return Response(202, {"Location": self._loc(s),
                                  "Docker-Upload-UUID": s.uuid}, None)
        s = UploadSession(name)
        self.uploads[s.uuid] = s
        return Response(202, {"Location": self._loc(s),
                              "Docker-Upload-UUID": s.uuid,
                              "Range": "0-0"}, None)

    @staticmethod
    def _loc(s):
        return "/v2/%s/blobs/uploads/%s" % (s.name, s.uuid)

    def _session(self, uuid_):
        if uuid_ not in self.uploads:
            raise RegistryError(404, "BLOB_UPLOAD_UNKNOWN",
                                "blob upload unknown to registry")
        return self.uploads[uuid_]

    def patch_upload(self, uuid_, data, content_range=None):
        s = self._session(uuid_)
        if content_range is not None:
            if not CONTENT_RANGE_RE.match(content_range):
                raise RegistryError(416, "BLOB_UPLOAD_INVALID",
                                    "bad content range %r" % content_range)
            start = int(content_range.split("-")[0])
            if s.offset == 0 and start != 0:
                raise RegistryError(416, "BLOB_UPLOAD_INVALID",
                                    "the first chunk's range must begin with 0")
            if start != s.offset:
                raise RegistryError(416, "BLOB_UPLOAD_INVALID",
                                    "chunk out of order: expected %d got %d"
                                    % (s.offset, start))
            if len(data) != int(content_range.split("-")[1]) - start + 1:
                raise RegistryError(416, "BLOB_UPLOAD_INVALID",
                                    "content-length does not match range")
        s.data += data
        return Response(202, {"Location": self._loc(s),
                              "Range": "0-%d" % (s.offset - 1),
                              "Docker-Upload-UUID": s.uuid}, None)

    def put_upload(self, uuid_, digest, data=None, content_range=None):
        s = self._session(uuid_)
        if data:
            self.patch_upload(uuid_, data, content_range)
        got = digest_of(s.data)
        if got != digest:
            raise RegistryError(400, "DIGEST_INVALID",
                                "provided digest did not match uploaded content",
                                {"provided": digest, "computed": got})
        repo = self.ensure_repo(s.name)
        repo["blobs"][got] = s.data
        del self.uploads[uuid_]
        return Response(201, {"Location": "/v2/%s/blobs/%s" % (s.name, got),
                              "Docker-Content-Digest": got}, None)

    def get_upload_status(self, uuid_):
        s = self._session(uuid_)
        return Response(204, {"Location": self._loc(s),
                              "Range": "0-%d" % (s.offset - 1 if s.offset else 0)}, None)

    def delete_upload(self, uuid_):
        self._session(uuid_)
        del self.uploads[uuid_]
        return Response(204, {}, None)

    # ---------------------------------------------------------------- end-8
    def list_tags(self, name, n=None, last=None):
        repo = self.repo(name)
        tags = sorted(repo["tags"].keys())
        if last is not None:
            tags = [t for t in tags if t > last]
        headers = {}
        if n is None:
            page, rest = tags, []
        else:
            page, rest = tags[:n], tags[n:]
        if rest and n != 0:
            headers["Link"] = ('</v2/%s/tags/list?n=%d&last=%s>; rel="next"'
                               % (name, len(rest), page[-1] if page else ""))
        return Response(200, headers, {"name": name, "tags": page})

    # ---------------------------------------------------------------- end-9/10
    def delete_manifest(self, name, reference):
        repo = self.repo(name)
        if not self.delete_enabled:
            raise RegistryError(405, "UNSUPPORTED", "the operation is unsupported")
        digest = repo["tags"].get(reference, reference)
        if digest not in repo["manifests"]:
            raise RegistryError(404, "MANIFEST_UNKNOWN", "manifest unknown to registry")
        del repo["manifests"][digest]
        for tag, d in list(repo["tags"].items()):
            if d == digest:
                del repo["tags"][tag]
        return Response(202, {}, None)

    def delete_blob(self, name, digest):
        repo = self.repo(name)
        if not self.delete_enabled:
            raise RegistryError(405, "UNSUPPORTED", "the operation is unsupported")
        if digest not in repo["blobs"]:
            raise RegistryError(404, "BLOB_UNKNOWN", "blob unknown to registry")
        del repo["blobs"][digest]
        return Response(202, {}, None)

    # ---------------------------------------------------------------- end-12
    def get_referrers(self, name, digest, artifact_type=None):
        repo = self.repo(name)
        if not self.referrers_enabled:
            # 不支持该 API 的 registry 回落到 referrers tag schema，客户端按 404 处理
            raise RegistryError(404, "UNSUPPORTED", "referrers API not supported")
        if not digest.startswith("sha256:") or len(digest) != 71:
            raise RegistryError(400, "UNSUPPORTED", "invalid digest syntax")
        digests = repo["referrers"].get(digest, [])
        headers = {"Content-Type": INDEX_MT}
        manifests = []
        for d in digests:
            m = repo["manifests"][d]
            desc = {"mediaType": m.get("mediaType", MANIFEST_MT), "digest": d,
                    "artifactType": m.get("artifactType",
                                          m.get("config", {}).get("mediaType", "")),
                    "annotations": m.get("annotations", {})}
            if artifact_type and desc["artifactType"] != artifact_type:
                continue
            manifests.append(desc)
        if artifact_type:
            headers["OCI-Filters-Applied"] = "artifactType"
        return Response(200, headers,
                        {"schemaVersion": 2, "mediaType": INDEX_MT,
                         "manifests": manifests})
